toolchain/check/member_access.cpp

```diff
@@ -81,15 +81,19 @@ static auto GetHighestAllowedAccess(Context& context, SemIR::LocId loc_id,
 
     // If the `type_id` of `Self` does not match with the one we're currently
     // accessing, try checking if this class is of the parent type of `Self`.
-    if (auto base_decl = context.insts().TryGetAsIfValid<SemIR::BaseDecl>(
-            self_class_info.base_id)) {
-      if (base_decl->base_type_id == class_info.self_type_id) {
+    if (auto base_type_id = self_class_info.GetBaseType(
+            context.sem_ir(), self_class_type->specific_id);
+        base_type_id.is_valid()) {
+      if (context.types().GetConstantId(base_type_id) == name_scope_const_id) {
         return SemIR::AccessKind::Protected;
       }
-    } else if (auto adapt_decl =
-                   context.insts().TryGetAsIfValid<SemIR::AdaptDecl>(
-                       self_class_info.adapt_id)) {
-      if (adapt_decl->adapted_type_id == class_info.self_type_id) {
+      // TODO: Also check whether this base class has a base class of its own.
+    } else if (auto adapt_type_id = self_class_info.GetAdaptedType(
+                   context.sem_ir(), self_class_type->specific_id);
+               adapt_type_id.is_valid()) {
```

Comment:
You've written similar in a couple spots (e.g. context.cpp), where the same logic is done on the extended type. Would it be worth having a function like `GetExtendedType` that returns `base` _or_ `adapt`?

Reply:
I plan to send a follow-up PR that removes the adapter handling here and in `DiagnoseInvalidQualifiedNameAccess`, because it doesn't match the design (see #4560), which I think would not leave much use of this facility.